nemo_rl/environments/genrm_environment.py

```
@@ -0,0 +1,277 @@
+        """Evaluate GenRM predictions and return rewards."""
+        
+        rewards = []
+        observations = []
+        
+        for conversation, meta in zip(message_log_batch, metadata):
+            # Extract assistant's response
+            assistant_response = ""
+            for msg in conversation:
+                if msg["role"] == "assistant":
+                    assistant_response = msg["content"]
+                    break
+            
+            
+            
+            is_valid_format, individual_scores, preference_ranking, error_message = self.validate_format_and_extract(
+                assistant_response, meta["num_responses"]
+            )
+            # Validate format and extract scores
+            print("assistant_response: ", assistant_response)
+            print("extracted results: ")
+            if individual_scores is not None:
+                print(individual_scores)
+            if preference_ranking is not None:
+                print(preference_ranking)
+            print()
+            
+            
+            if not is_valid_format:
+                # Apply format penalty
+                reward = self.format_penalty
+                rewards.append(float(reward))
+                observations.append({
+                    "role": "environment",
+                    "content": f"Format violation penalty applied. Error: {error_message}. Reward: {reward}"
+                })
+                
+            else:
+                reward = self.format_penalty
+                try:
+                    if meta["num_responses"] == 1:
+                        if meta["helpfulness_1"] is not None and individual_scores:
+                            reward = -self.distance_abs(individual_scores[0], meta["helpfulness_1"])
+                    
+                    elif meta["num_responses"] == 2:
```

Safer to add also an `else:` block with an exception just to be sure there's no silent failure with >2 responses